### PR TITLE
Webpack: Prevent long preview output filenames

### LINKED ISSUE #30221: [Bug]: ENAMETOOLONG with deeply nested folder structure in storybook/next
### Describe the bug

If you have a folder structure with long names that is deeply nested, it's possible to overrun the OS's maximum file length as there is no built in truncation

You should then see an issue like

```
=> Failed to build the preview
SB_BUILDER-WEBPACK5_0002 (WebpackInvocationError): ENAMETOOLONG: name too long, open './storybook-static/app-some-folder2-some-sub-folder-some-sub-sub-folder-some-sub-sub-sub-folder-some-sub-sub-sub-sub-folder-some-sub-sub-sub-sub-sub-folder-some-sub-sub-sub-sub-sub-sub-folder-something_is_going_to_get_cha-SomeComponent-AComponent-stories.a38fd5fb.iframe.bundle.js'
```
This error is because the file name being written to the disk is too long and the OS rejects it. It appears as though the config is using a concatenation of the entire chunk file name and path to create the output file name. This error doesn't happen in standard webpack or in next.js, they have a maxlength for the filename and truncate it.

webpack for example
```
dist/src_some-folder_some-sub-folder_some-sub-sub-folder_some-sub-sub-sub-folder_some-sub-sub-sub--921ea3.js
```

next.js for example
```
.next/static/chunks/9e8ad_something_is_going_to_get_cha_SomeComponent_AComponent_tsx_2394de._.js
```

it is possible to work around this issue by manually setting the webpack config output.filename to not include the [name] param.

### Reproduction link

https://github.com/diffidentDude/next-test

### Reproduction steps

I have created a repo to repro the issue https://github.com/diffidentDude/next-test

### System

```bash
% npx storybook@latest info
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm warn deprecated rimraf@2.6.3: Rimraf versions prior to v4 are no longer supported

Storybook Environment Info:

  System:
    OS: macOS 14.7
    CPU: (10) arm64 Apple M1 Pro
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.14.0 - ~/Code/next-test/.devbox/nix/profile/default/bin/node
    npm: 10.7.0 - ~/Code/next-test/.devbox/nix/profile/default/bin/npm <----- active
    pnpm: 9.11.0 - ~/Code/next-test/.devbox/virtenv/nodejs/corepack-bin/pnpm
  Browsers:
    Chrome: 131.0.6778.205
    Safari: 18.0
  npmPackages:
    @storybook/addon-essentials: ^8.4.7 => 8.4.7 
    @storybook/addon-interactions: ^8.4.7 => 8.4.7 
    @storybook/addon-onboarding: ^8.4.7 => 8.4.7 
    @storybook/blocks: ^8.4.7 => 8.4.7 
    @storybook/nextjs: ^8.4.7 => 8.4.7 
    @storybook/react: ^8.4.7 => 8.4.7 
    @storybook/test: ^8.4.7 => 8.4.7 
    eslint-plugin-storybook: ^0.11.2 => 0.11.2 
    storybook: ^8.4.7 => 8.4.7
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by zhangli091011**
Closes #30221

Claim: https://github.com/storybookjs/storybook/issues/30221#issuecomment-5018627579

## What I did

- Replaced the webpack5 preview output `[name]` templates with a filename function that preserves normal readable names.
- Bounded long entry and async chunk filename components to 200 bytes, using a sanitized prefix and deterministic 64-bit SHA-256 suffix to avoid collisions.
- Preserved production `[contenthash:8]` behavior and applied the same strategy in development.
- Added focused helper tests and real in-memory webpack compilations covering development and production entry/chunk output.

AI disclosure: GPT-5.6 via OpenCode assisted with implementation and validation. I reviewed the code, tests, and PR content.

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [x] unit tests
- [x] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Create a webpack5 Storybook whose stories are under a deeply nested path long enough to previously produce `ENAMETOOLONG`.
2. Run the Storybook development server and confirm the preview compiles without `ENAMETOOLONG`.
3. Run a static Storybook build and confirm it succeeds.
4. Inspect the output and confirm normal chunk names remain readable, while long entry and async chunk filename components are at most 200 bytes and production files retain an 8-character content hash.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Declare whether manual QA will be needed for this PR during the next release, through `qa:needed` or `qa:skip`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

   - `bug`: Internal changes that fixes incorrect behavior.
   - `maintenance`: User-facing maintenance tasks.
   - `dependencies`: Upgrading (sometimes downgrading) dependencies.
   - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
   - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
   - `documentation`: Documentation **only** changes. Will not show up in release changelog.
   - `feature request`: Introducing a new feature.
   - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
   - `other`: Changes that do not fit in the above categories.

   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=<PR_NUMBER>`_

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved preview iframe chunk filename generation for more consistent, readable output in development and the expected placeholder behavior in production.
  * Ensures filenames are filesystem-safe, deterministic, and unique for edge-case or very long chunk names, with a strict byte-size limit.
* **Tests**
  * Added unit and in-memory integration-style coverage for preview filename generation, sanitization, hashing/uniqueness, and emitted asset naming.
* **Chores**
  * Added `memfs` to the test setup for the preview filename environment.

### CODE REVIEW

File: code/builders/builder-webpack5/src/preview/preview-filename.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      return `${name}${suffix}`;
+    }
+
+    const hashSuffix = `-${hash(name)}`;
+    const maxNameBytes = MAX_FILENAME_LENGTH - Buffer.byteLength(hashSuffix + suffix);
+    const prefix = truncateToBytes(sanitizedName, maxNameBytes).replace(/[ .]+$/g, '') || 'chunk';
+
+    return `${prefix}${hashSuffix}${suffix}`;
+  };
+}
```
[the remark this answers is not in the document]
Comment (zhangli091011) [reply code/builders/builder-webpack5/src/preview/preview-filename.ts:46]: Fixed in 96ef5fc8b9c. Empty `chunk.name` now falls back to `chunk.id`, numeric ID `0` is retained, and `chunk` is used only when neither is available. Added focused tests for an empty name with an ID, empty/missing names without an ID, and ID `0`. Validation passed: 23 builder-webpack5 tests, `yarn nx check builder-webpack5`, `yarn nx compile builder-webpack5`, formatting, and focused lint.

**Comment by zhangli091011** [comment]
When convenient, could a maintainer please apply the appropriate labels? `bug`, `ci:normal`, and `qa:needed` seem suitable for this webpack output fix, though please use the QA label that best matches the project workflow. Contributor permissions do not allow me to add them.

**Comment by ndelangen** [comment]
@zhangli091011 Thank you so much for this PR!

Could we make this fix only apply to overlong async chunk filenames? The current implementation also changes normal filenames and short path-like chunk names, which seems broader than the reported bug.

**Comment by zhangli091011** [comment]
Updated in `daa0ca3e11f` to narrow the change as requested. The original `output.filename` templates are restored, short async chunk names (including path-like names) are returned unchanged, and only async chunk filenames over the 200-byte threshold are sanitized, truncated, and given a deterministic hash. I also adjusted the in-memory webpack tests to verify normal entry filenames remain unchanged in development and production while long async chunks are bounded. Validation passed: 10 focused tests, `yarn nx check builder-webpack5`, formatting, and focused lint.

**Comment by zhangli091011** [comment]
The branch is now updated to the latest `next` (`3d78a5e7cac`). The narrowed implementation remains conflict-free, but CircleCI is still skipped because the required change, CI, and QA labels are unavailable to contributors. Could a maintainer please apply the appropriate labels (for example `bug`, `ci:normal`, and the suitable `qa:*` label) and re-review the scoped fix?

**Comment by zhangli091011** [comment]
Merged the latest `next` in `06c88f968a0` and resolved the lockfile conflict without widening the fix: normal output filenames and short/path-like async chunk names remain unchanged, only overlong async chunk filenames are bounded, and empty names still fall back correctly. Validation passed: all 23 builder-webpack5 tests (including in-memory development/production webpack builds), `yarn nx check builder-webpack5`, formatting, and focused lint (0 errors). Could a maintainer please add the required `ci:*` and `qa:*` labels and re-review?

**Review by ndelangen — APPROVED** [review]
